src/main/kotlin/ar/edu/unahur/obj2/caralibro/Usuario.kt

```diff
@@ -1,11 +1,55 @@
 package ar.edu.unahur.obj2.caralibro
 
 class Usuario {
-  val publicaciones = mutableListOf<Publicacion>()
+  private var publicaciones = mutableListOf<Publicacion>()
+
+  private var amigos = mutableListOf<Usuario>()
 
   fun agregarPublicacion(publicacion: Publicacion) {
     publicaciones.add(publicacion)
+    publicacion.propietario = this
   }
 
   fun espacioDePublicaciones() = publicaciones.sumBy { it.espacioQueOcupa() }
+
+  fun puedeVerPublicacionDe(publicacion: Publicacion) : Boolean {
+    return (publicacion.permisos == Permisos.PUBLICA
+            || (publicacion.permisos == Permisos.EXCLUIDOS && !publicacion.listaExcluidos.contains(this))
+            || (publicacion.permisos == Permisos.PERMITIDOS && publicacion.listaPermitidos.contains(this))
+            || (publicacion.permisos == Permisos.PRIVADA && publicacion.propietario?.amigos?.contains(this) ?: false))
+  }
+
+  fun agregarAmigo(usuarioAmigo: Usuario) {
+    amigos.add(usuarioAmigo)
+  }
+
+  fun masAmistosoQue(otroUsuario : Usuario) = this.amigos.size > otroUsuario.amigos.size
+
+  fun listaMejoresAmigos() : MutableList<Usuario> {
+    var listaMejoresAmigos = this.amigos
+    val listaPersonasExcluidas = listaTotalDePersonasExcluidas()
+    for (personaExcluida in listaPersonasExcluidas){
+      if (listaMejoresAmigos.contains(personaExcluida)){
+        listaMejoresAmigos.remove(personaExcluida)
+      }
+    }
+    return listaMejoresAmigos
+  }
+
+   private fun listaTotalDePersonasExcluidas() : MutableList<Usuario> {
+    var listaExcluidos = mutableListOf<Usuario>()
+    for(publicaciones in this.publicaciones){
+      for (excluidos in publicaciones.listaExcluidos){
+        listaExcluidos.add(excluidos)
+      }
+    }
+    return listaExcluidos
```

Comment:
Bueno, ya dije que esta lógica estaba mal, pero les comento algo respecto al `for`:

En Kotlin, no necesitás usar el `for` casi para nada. Siempre que escriban un for, frenen y fíjense en la documentación (o nos preguntan) si existe algo para resolver ese problema.

En este caso, existe [flatMap](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/flat-map.html) que les permite reescribir esto como:

```kotlin
fun listaTotalDePersonasExcluidas() = publicaciones.flatMap { it.listaExcluidos }
```